design/namespaced-roles.md

```diff
@@ -0,0 +1,299 @@
+   - Supports empty values - A role might be created and assigned at a realm level
+   - Example: `role_v1:/realm1/tenantA///roleName`, `role_v1:/realm1///groupA/roleName`
+
+**Client**: Roles can be defined and assigned for a certain client in Keycloak, so a unique identifier has to support this possibility.
+   - Supports empty values - A role might be created and assigned without being restricted to a client.
+   - Example: `role_v1:/realm1/tenantA/clientId//roleName`, `role_v1:/realm1/tenantA//groupA/roleName`
+
+**Group**: Roles can be assigned to groups or, as intended in this design proposal, assigned to a user in the context of a group.
+   - Supports empty values - A role might be created at any level and not necessarily assigned to a group.
+   - Example: `role_v1:/realm1/tenantA/clientId/groupB/roleName`, `role_v1:/realm1/tenantA/clientId//roleName`
+
+**RoleName**: Finally, the name of the role that is being defined.
+   - Doesn't support empty values - You need to know the name of the role that is being created.
+   - Example: `role_v1:/realm1/tenantA/clientId/groupB/roleName`
+
+A possible free-form role identifier could be allowed for resource-server specific roles.
+
+The initial fixed part would be kept in order to identify the role within a realm, tenant, client, group etc, but instead of a fixed name for the role, it would be a reference to other hierarchical role formed with any random structure.
+
+A possible format would be:
+
+- `role_v1:/realm1/tenant1//group1/frole_v1:/mycompany/resources/department-a-roles/developer`
+
+If the role name is another reference to a free form role, starting with `frole_v1:/`, this role will also be represented in a hierarchical way, making it possible for protocol mappers to show them as such in tokens.
+
+## Use Cases
+
+As previously mentioned, there are a few use cases that we'll want to support and are currently not possible in Keycloak.
+
+### Use Case 1 - Assign a role in the context of a group
+
+Say we want to have the following group composition in our realm (or our tenant, in the future), most likely imported from a federated LDAP:
+
+realm/tenant:tenant1
+ - group:iam
+ - group:devops
+ - group:leadership
+
+Following the current implementation, the users would need to define roles at the realm level and then assign them to these groups. Everyone in those groups would inherit the roles.
+
+But for a client, these groups may have business meaning or reflect organizational structure within their company, they aren't just a mean to group users together with the same roles, but rather fully functional teams that users can be part of with different roles.
+
+So UserA would be a `manager` in `group:iam`, but UserB could be a `developer`in `group:iam`.
+
+Namespaced roles would be specified when created, and loosely assigned based on the namespace, so let's say we have the following roles:
+
+`role_v1:/realm1/tenant1//iam/manager`
```

Comment:
I don't understand the meaning of the double forward slash (//). I also somewhat think that this is being complicated by including the tenant in the namespace. Perhaps a simpler example for role within a group is just "role_v1:/iam/manager'

Reply:
The double slashes are basically a missing field.

Take `role_v1:/tenant1/client1/iam/manager`, that contains values for all fields, because it's created in the scope of a client.

Now, imagine you wanted to create the same role, but without the client, in order to keep the order of variables and make it possible to retrieve by position, you need to keep an empty space between tenant and group, hence:

`role_v1:/tenant//iam/manager`.

Reply:
What I had in mind wasn't to have some special meaning of the namespace, but rather it's just a free-form hierarchy. So you would just use `role_v1:/tenant/iam/manager` or `role_v1:/iam/manager`. Up to you. I was thinking we'd have a controlled namespace for internal roles like `role_v1:/kc/clients/{client-name}/{role-name}`